# Project 2: Oil and Gas Summary Production Data (1967-1999)


####Data source site: https://catalog.data.gov/dataset?res_format=CSV
####Datasheet link: https://catalog.data.gov/dataset/oil-and-gas-summary-production-data-1967-1999

In [4]:
import pandas as pd
import numpy as np
from altair import Chart, X, Y, Color, Scale
import altair as alt

In [5]:
production_df = pd.read_csv('/content/drive/My Drive/HDS/CMPS 260/Data/Oil_and_Gas_Summary_Production_Data__1967-1999.csv')
production_df.head()

Production Year  ...                                  Location
0             1995  ...      BUFFALO, NY\n(42.887691, -78.879374)
1             1995  ...   WEST UNION, NY\n(42.094951, -77.725816)
2             1994  ...   CARROLLTON, NY\n(42.023289, -78.628438)
3             1994  ...  CHEEKTOWAGA, NY\n(42.907071, -78.754318)
4             1995  ...     ALLEGANY, NY\n(42.088061, -78.491258)

[5 rows x 20 columns]

In [6]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


###Data Column meanings: 

1. **Production Year**: The year during which the oil and gas are produced. 
2. **Production date entered**: The date of production inputted by the operator.
3.**Operator**: The company or individual that are responsible and have the legal authority to drill wells and undertake the production of hydrocarbons that are found.
4. **County**: Specific region of a state or country where the oil and gas wells are located at.
5. **Town**: The town where the oil and gas wells are located at. 
6. **Field**: The geographical area under which an oil or gas reservoir lies.
7. **Producing formation**: The area where there is an underground rock formation from which oil, gas or water is produced
8. **Active oil wells**: Number of wells that are currently producing oil. 
9. **Inactive oil wells**: Number of wells that have not produced oil in 12 months.
10. **Active gas wells**: Number of wells that are currently producing gas. 
11. **Inactive gas wells**: Number of wells that have not produced gas in 12 months.  
12. **Injection wells**: An injection well is used to place fluid underground into porous geologic formations. This could be steam, carbon dioxide, water or other substances to help maintain reservoir pressure, heat the oil or lower its viscosity, allowing it to flow to a producing well nearby.
13. **Disposal wells**: Number of depleted oil or gas wells, into which waste fluids can be injected for safe disposal (extra info: Water specific to fracture return may be recycled for further use, or disposed by injection into deep subsurface formations, through a disposal well.)
14. **Self-use well**:
15. **Oil produced (bbl)**: Total number of barrels of oil produced by the active oil wells. 
16. **Gas produced (mcf)**: Amount of gas, in Mille Cubic Feet or 1000 cubic feet,  produced by active gas wells.
17. **Water produced (bbl)**: Total number of barrels of water produced by the active oil and gas wells.
18. **Taxable gas (mcf)**: Amount of gas, in Mille Cubic Feet or 1000 cubic feet, that can be taxed. 
19. **Purchaser codes**: Specifies a code to specify the purchaser who normally handles this vendor's account.
20. **Location**: The country, state, and longitude and latitude 



###Cleaning the Data:

####First let's check which columns have empty observations. 

In [7]:
production_df.shape

(30053, 20)

In [8]:
production_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30053 entries, 0 to 30052
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Production Year          30053 non-null  int64 
 1   Production Date Entered  30053 non-null  object
 2   Operator                 30053 non-null  object
 3   County                   30022 non-null  object
 4   Town                     29396 non-null  object
 5   Field                    28772 non-null  object
 6   Producing Formation      29393 non-null  object
 7   Active Oil Wells         30053 non-null  int64 
 8   Inactive Oil Wells       30053 non-null  int64 
 9   Active Gas Wells         30053 non-null  int64 
 10  Inactive Gas Wells       30053 non-null  int64 
 11  Injection Wells          30053 non-null  int64 
 12  Disposal Wells           30053 non-null  int64 
 13  Self-use Well            29434 non-null  object
 14  Oil Produced, bbl        30053 non-nul

####Since there are supposed to be 30,053 observations in total, we can see that the County, Town, Field, Producing Formation, Self-Use Well, and Purchaser Codes columns have empty rows. We will now remove these empty rows. We will not bother about the Self-Use Well and Purchaser Codes columns since those are not used in our analysis. 


In [9]:
production_df = production_df.loc[production_df['County'].notnull()]
production_df = production_df.loc[production_df['Town'].notnull()]
production_df = production_df.loc[production_df['Field'].notnull()]
production_df = production_df.loc[production_df['Producing Formation'].notnull()]
production_df.shape


(28357, 20)

####Since the observations have decreased from 30,053 to 28,357, the null rows have been removed. 

##**Analysis Questions:**


---


###**1) The Renewable Energy movement started during the 1970s. Assuming that this movement would decrease the dependency on natural gas and oil by making people more aware, do the number of active  oil and gas wells decrease over the years? Can we also see the same effect on the number of disposal wells?**


---













#### First, we will make a new dataframe `NR_df` which only has the columns that we need for this analysis. 

In [10]:
NR_df = production_df[['Production Year', 'Active Oil Wells', 'Active Gas Wells', 'Disposal Wells']]
NR_df.head()

Production Year  Active Oil Wells  Active Gas Wells  Disposal Wells
0             1995                 0                 1               0
1             1995                28                 0               0
2             1994                 3                 0               0
3             1994                 0                 1               0
4             1995                10                 0               0

#### We will now group by the production year to see the counts of the 5 different types of wells in each year.

In [11]:
NR_df1 = NR_df.groupby('Production Year').sum()
NR_df1.head()

Active Oil Wells  Active Gas Wells  Disposal Wells
Production Year                                                    
1967                         3250               484               0
1968                         3678               515               0
1969                         4117               481               0
1970                         4261               446               0
1971                         4229               409               0

####Using the `melt()` function we will turn this dataframe from a wide form to a long form since Altair charts work best with the long form.

In [12]:
NR_df1 = NR_df1.reset_index()
NR_df1.head()

Production Year  Active Oil Wells  Active Gas Wells  Disposal Wells
0             1967              3250               484               0
1             1968              3678               515               0
2             1969              4117               481               0
3             1970              4261               446               0
4             1971              4229               409               0

In [13]:
NR_df1_melt = NR_df1.melt(id_vars='Production Year', value_vars=['Active Oil Wells', 'Active Gas Wells', 'Disposal Wells'])
NR_df1_melt

Production Year          variable  value
0              1967  Active Oil Wells   3250
1              1968  Active Oil Wells   3678
2              1969  Active Oil Wells   4117
3              1970  Active Oil Wells   4261
4              1971  Active Oil Wells   4229
..              ...               ...    ...
94             1995    Disposal Wells      0
95             1996    Disposal Wells      1
96             1997    Disposal Wells      5
97             1998    Disposal Wells      0
98             1999    Disposal Wells      0

[99 rows x 3 columns]

####We are now representing this melted dataframe using a line chart. 

In [14]:
Chart(NR_df1_melt).mark_line().encode(x='Production Year:O', y='value', color='variable')

alt.Chart(...)

#### Unfortunately, from the line chart above, we can see that the Renewable Energy movement did not decrease the number of active oil wells and gas wells. Infact, there is an increase in active gas wells significantly from 1975 which is contradictory to the purpose of the movement. We do see a decline in number of active oil wells from 1978 which might indicate more dependency on renewable sources. However, we end up having more number of active oil wells in 1999 than in 1967. For the disposal wells, there seems to be no significant change except for the brief increases in number during 1978, 1987 and 1992. 



---


###**2) Injection wells are said to improve oil and gas production. Can we see a positive relationship between number of injection wells and amount of gas and oil produced?**

---







####First, we will group our data by injection wells and bring out only the columns that we need which are Injection Wells, Oil Produced and Gas Produced.

In [15]:
IW_df = production_df[['Injection Wells', 'Oil Produced, bbl','Gas Produced, Mcf']]
IW_df = IW_df.groupby('Injection Wells').sum()
IW_df = IW_df.reset_index()
IW_df.head()


Injection Wells  Oil Produced, bbl  Gas Produced, Mcf
0                0           16613675          515658270
1                1              43017            1330659
2                2              36288              22390
3                3              52766               9266
4                4              13850               4513

###Now, we will create a new column called 'Injection Wells Category'. This column will consist information about whether the number of injection wells is Low, Moderate or High. To do this I will first create a function called IW_category.

In [16]:
IW_df['Injection Wells'].describe()

count     150.000000
mean      136.880000
std       161.157263
min         0.000000
25%        37.250000
50%        81.500000
75%       178.750000
max      1108.000000
Name: Injection Wells, dtype: float64

#### To categorize the number of injection wells into 'Low Injection Well Count', 'Moderate Injection Well Count', and 'High Injection Well Count', we will use the first quadrant and third quadrant values as boundaries. 

In [17]:
def IW_category(Injection_Wells):
  if Injection_Wells <= 37.25:
    return 'Low Injection Well Count'
  elif Injection_Wells >= 178.75:
    return 'High Injection Well Count'
  else:
    return 'Moderate Injection Well Count'

#### Now we will apply `IW_category` function to the "Injection Wells" column to create the new category column.

In [18]:
IW_df['Injection Wells Category'] = IW_df['Injection Wells'].apply(IW_category)
IW_df.head()

Injection Wells  ...  Injection Wells Category
0                0  ...  Low Injection Well Count
1                1  ...  Low Injection Well Count
2                2  ...  Low Injection Well Count
3                3  ...  Low Injection Well Count
4                4  ...  Low Injection Well Count

[5 rows x 4 columns]

###Using new method of creating different tables for different injection wells *category*

In [19]:
IW_low = IW_df.loc[IW_df['Injection Wells Category'] == 'Low Injection Well Count']
IW_low.head()

Injection Wells  ...  Injection Wells Category
0                0  ...  Low Injection Well Count
1                1  ...  Low Injection Well Count
2                2  ...  Low Injection Well Count
3                3  ...  Low Injection Well Count
4                4  ...  Low Injection Well Count

[5 rows x 4 columns]

#### Grouping by the Injection Wells Category:

In [20]:
IW_df1 = IW_df.groupby('Injection Wells Category')[['Oil Produced, bbl', 'Gas Produced, Mcf']].median()
IW_df1.head()

Oil Produced, bbl  Gas Produced, Mcf
Injection Wells Category                                           
High Injection Well Count                  66346                  0
Low Injection Well Count                   15938               2918
Moderate Injection Well Count              13616                  0

In [21]:
IW_df1 = IW_df1.reset_index()
IW_df1

Injection Wells Category  Oil Produced, bbl  Gas Produced, Mcf
0      High Injection Well Count              66346                  0
1       Low Injection Well Count              15938               2918
2  Moderate Injection Well Count              13616                  0

####Since these values are really big, we will take the natural log of these values. 

In [22]:
IW_df1['Natural log of Oil Produced, bbl'] = np.log(IW_df1['Oil Produced, bbl'])
IW_df1['Natural log of Gas Produced, Mcf'] = np.log(IW_df1['Gas Produced, Mcf'])
IW_df1 = IW_df1[['Injection Wells Category', 'Natural log of Gas Produced, Mcf', 'Natural log of Oil Produced, bbl']]
IW_df1

/usr/local/lib/python3.7/dist-packages/pandas/core/series.py:726: RuntimeWarning: divide by zero encountered in log
  result = getattr(ufunc, method)(*inputs, **kwargs)


Injection Wells Category  ...  Natural log of Oil Produced, bbl
0      High Injection Well Count  ...                         11.102639
1       Low Injection Well Count  ...                          9.676461
2  Moderate Injection Well Count  ...                          9.519001

[3 rows x 3 columns]

In [23]:
Chart(IW_df1).mark_bar().encode(x='Natural log of Gas Produced, Mcf', y='Injection Wells Category')

alt.Chart(...)

In [24]:
Chart(IW_df1).mark_bar().encode(x='Natural log of Oil Produced, bbl', y='Injection Wells Category')

alt.Chart(...)

**3) It is stated Oil wells sometimes produce large volumes of water with the oil, while gas wells tend to produce water in smaller proportions. Do the operator, field, or producing formation producing higher amounts of water have a higher proportion of active oil wells?
-create new column active oil wells>active gas wells.**



Grouping

In [25]:
WP_df = production_df[['Production Year', 'Active Oil Wells', 'Active Gas Wells', 'Oil Produced, bbl', 'Gas Produced, Mcf', 'Water produced, bbl']]
WP_df.head()

Production Year  Active Oil Wells  ...  Gas Produced, Mcf  Water produced, bbl
0             1995                 0  ...                106                    0
1             1995                28  ...                  0                  180
2             1994                 3  ...                  0                    0
3             1994                 0  ...                530                    0
4             1995                10  ...                  0                   80

[5 rows x 6 columns]

In [26]:
WP_df['Water produced, bbl'].describe()

count    2.835700e+04
mean     6.194496e+03
std      1.766145e+05
min      0.000000e+00
25%      0.000000e+00
50%      0.000000e+00
75%      2.000000e+00
max      1.759123e+07
Name: Water produced, bbl, dtype: float64

In [27]:
WP_df['Gas Produced, Mcf'].describe()

count    2.835700e+04
mean     1.831010e+04
std      8.411717e+04
min      0.000000e+00
25%      0.000000e+00
50%      4.450000e+02
75%      7.706000e+03
max      4.179174e+06
Name: Gas Produced, Mcf, dtype: float64

In [28]:
WP_df['Oil Produced, bbl'].describe()

count     28357.000000
mean        833.457912
std        8478.167001
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max      675589.000000
Name: Oil Produced, bbl, dtype: float64

In [29]:
production_types = {'Oil production', 'Gas production'}
production_types

{'Gas production', 'Oil production'}